### DISCUSSION TITLE
Way to integration with @AuthenticationPrincipal

### DISCUSSION THREAD

**Discussion by aisensiy** [Netflix/dgs-framework#126] [Q&A]
I just learn this framework for some days and I think this is such a great framework and I'd like to make my current [spring-boot-realworld-example-app](https://github.com/gothinkster/spring-boot-realworld-example-app) to support graphql with this framework.

But after learning the docs I still feel confusing about the authentication integration. I am using jwt as the token system and I implement a [JwtTokenFilter](https://github.com/gothinkster/spring-boot-realworld-example-app/blob/9306c39fcf31aa669f519b118dc724e811c37692/src/main/java/io/spring/api/security/JwtTokenFilter.java) to make the `@AuthenticationPrincipal` injection work in spring mvc controller:

```java
    @GetMapping
    public ResponseEntity<?> article(@PathVariable("slug") String slug,
                                     @AuthenticationPrincipal User user) {
        return articleQueryService.findBySlug(slug, user)
            .map(articleData -> ResponseEntity.ok(articleResponse(articleData)))
            .orElseThrow(ResourceNotFoundException::new);
    }
```

Is there some tips or resources for me to support `@AuthenticationPrincipal` or similar mechanism to easily access current user in every `Query` `Mutation` `Subscription`.

**Comment by berngp** [comment]
@aisensiy, the framework doesn't provide any opinions on how to solve AuthN, or AuthZ. In general, the [Spring Security docs](https://docs.spring.io/spring-security/site/docs/current/reference/html5/) offer better guidelines on how to achieve this.

Now, having stated that, I have not used the `AuthenticationPrincipal` annotation in this context but internally we use a variation of the `SecurityContextHolder`. You should be able to fetch the `Authentication` reference via the `SecurityContextHolder`, as below.

```
Authentication authentication = SecurityContextHolder.getContext().getAuthentication();
```

## Note

The `SecurityContextHolder` is dependent on _thread local_ variables and will not be accessible in a custom executor that is not decorated managed via Spring. For more information look into [Spring Security's TaskExecutor decoration](https://docs.spring.io/spring-security/site/docs/current/reference/html5/#concurrency).

**Comment by aisensiy** [comment]
Here is my update.

I create a utility method like this:

```java
public class SecurityUtil {
  public static Optional<User> getCurrentUser() {
    Authentication authentication = SecurityContextHolder.getContext().getAuthentication();
    if (authentication instanceof AnonymousAuthenticationToken
        || authentication.getPrincipal() == null) {
      return Optional.empty();
    }
    io.spring.core.user.User currentUser = (io.spring.core.user.User) authentication.getPrincipal();
    return Optional.of(currentUser);
  }
}
```

https://github.com/gothinkster/spring-boot-realworld-example-app/blob/master/src/main/java/io/spring/graphql/SecurityUtil.java

And using it like this:

```java
  @DgsData(parentType = MUTATION.TYPE_NAME, field = MUTATION.AddComment)
  public DataFetcherResult<CommentPayload> createComment(
      @InputArgument("slug") String slug, @InputArgument("body") String body) {
    User user = SecurityUtil.getCurrentUser().orElseThrow(AuthenticationException::new); // <--- CODE HERE
    Article article =
        articleRepository.findBySlug(slug).orElseThrow(ResourceNotFoundException::new);
    Comment comment = new Comment(body, user.getId(), article.getId());
    commentRepository.save(comment);
    CommentData commentData =
        commentQueryService
            .findById(comment.getId(), user)
            .orElseThrow(ResourceNotFoundException::new);
    return DataFetcherResult.<CommentPayload>newResult()
        .localContext(commentData)
        .data(CommentPayload.newBuilder().build())
        .build();
  }
```

https://github.com/gothinkster/spring-boot-realworld-example-app/blob/7d0edfd4d723e0b6221bc7e77d7939eb883a7076/src/main/java/io/spring/graphql/CommentMutation.java#L43

Is this the best way or is there some better way for this?
